apps/x509.c

```diff
@@ -67,97 +69,117 @@ typedef enum OPTION_choice {
 const OPTIONS x509_options[] = {
     OPT_SECTION("General"),
     {"help", OPT_HELP, '-', "Display this summary"},
-#ifndef OPENSSL_NO_ENGINE
-    {"engine", OPT_ENGINE, 's', "Use engine, possibly a hardware device"},
-#endif
 
-    {"inform", OPT_INFORM, 'f',
-     "CSR input format (DER or PEM) - default PEM"},
-    {"in", OPT_IN, '<', "Input file - default stdin"},
+    {"in", OPT_IN, '<',
+     "Certificate input (default stdin), or CSR input file with -req"},
     {"passin", OPT_PASSIN, 's', "Private key and cert file pass-phrase source"},
+    {"new", OPT_NEW, '-', "Generate a certificate from scratch"},
+    {"x509toreq", OPT_X509TOREQ, '-',
+     "Output a certification request (rather than a certificate)"},
+    {"req", OPT_REQ, '-', "Input is a CSR file (rather than a certificate)"},
+    {"inform", OPT_INFORM, 'f',
+     "CSR input file format (DER or PEM) - default PEM"},
+    {"vfyopt", OPT_VFYOPT, 's', "CSR verification parameter in n:v form"},
+    {"signkey", OPT_SIGNKEY, 's',
+     "Key used to self-sign certificate or cert request"},
+    {"keyform", OPT_KEYFORM, 'E',
+     "Key input format (ENGINE, other values ignored)"},
+    {"out", OPT_OUT, '>', "Output file - default stdout"},
     {"outform", OPT_OUTFORM, 'f',
      "Output format (DER or PEM) - default PEM"},
-    {"out", OPT_OUT, '>', "Output file - default stdout"},
-    {"keyform", OPT_KEYFORM, 'E', "Private key format (ENGINE, other values ignored)"},
-    {"req", OPT_REQ, '-', "Input is a certificate request, sign and output"},
-    {"vfyopt", OPT_VFYOPT, 's', "Verification parameter in n:v form"},
+    {"nocert", OPT_NOCERT, '-',
+     "No cert output (except for requested printing)"},
+    {"noout", OPT_NOOUT, '-', "No output (except for requested printing)"},
 
-    OPT_SECTION("Output"),
+    OPT_SECTION("Certificate printing"),
+    {"text", OPT_TEXT, '-', "Print the certificate in text form"},
+    {"certopt", OPT_CERTOPT, 's', "Various certificate text printing options"},
+    {"fingerprint", OPT_FINGERPRINT, '-', "Print the certificate fingerprint"},
+    {"alias", OPT_ALIAS, '-', "Print certificate alias"},
     {"serial", OPT_SERIAL, '-', "Print serial number value"},
-    {"subject_hash", OPT_HASH, '-', "Print subject hash value"},
-    {"issuer_hash", OPT_ISSUER_HASH, '-', "Print issuer hash value"},
-    {"hash", OPT_HASH, '-', "Synonym for -subject_hash"},
+    {"startdate", OPT_STARTDATE, '-', "Print the notBefore field"},
+    {"enddate", OPT_ENDDATE, '-', "Print the notAfter field"},
+    {"dates", OPT_DATES, '-', "Print both Before and After dates"},
```

Comment:
Should be `both notBefore and notAfter dates`?

Reply:
Good point, so renamed them for consistency with the cert field names
and mentioned this in `CHANGES.md`.
I think it's not worth adding option aliases for backward compatibility, is it?

Reply:
> I think it's not worth adding option aliases for backward compatibility, is it?

I think it's not worth, but I don't know if this would bring more inconvenient to the users.

And I think the text of option `OPT_DATES` should be `Print both notBefore and notAfter dates` since there are no fields called Before and After in an X.509 certificate ;-)

Reply:
I've revised the help and documentation texts for all three options.

Reply:
Having three options is kinda silly, unless you make notBefore and notAfter print out machine-readable, as in the ASN1 string?

Reply:
> Having three options is kinda silly

I wonder if it's the right time to discuss such details on rather irrelevant options.
Let's focus now on getting the essence of the apps usable and well documented.

Reply:
> I wonder if it's the right time to discuss such details on rather irrelevant options.
> Let's focus now on getting the essence of the apps usable and well documented.

I wonder if now is the right time to add new names to existing options.